src/Searching.js

```diff
@@ -98,14 +146,318 @@ async function localSearch(searchTerm, roomId = undefined) {
+        };
+
+        return MatrixClientPeg.get()._processRoomEventsSearch(emptyResult, response);
+    }
+
+    searchArgs.next_batch = localResult.next_batch;
+
+    const result = {
+        response: localResult,
+        query: searchArgs,
+    };
+
+    return result;
+}
+
+async function localPagination(searchResult) {
+    const eventIndex = EventIndexPeg.get();
+
+    const searchArgs = searchResult.seshatQuery;
+
+    const localResult = await eventIndex.search(searchArgs);
+    searchResult.seshatQuery.next_batch = localResult.next_batch;
+
     const response = {
         search_categories: {
             room_events: localResult,
         },
     };
 
-    const result = MatrixClientPeg.get()._processRoomEventsSearch(
-        emptyResult, response);
+    const result = MatrixClientPeg.get()._processRoomEventsSearch(searchResult, response);
+    searchResult.pendingRequest = null;
+
+    return result;
+}
+
+function compareOldestEvents(firstResults, secondResults) {
+    try {
+        const oldestFirstEvent = firstResults.results[firstResults.results.length - 1].result;
+        const oldestSecondEvent = secondResults.results[secondResults.results.length - 1].result;
+
+        if (oldestFirstEvent.origin_server_ts <= oldestSecondEvent.origin_server_ts) {
+            return -1;
+        } else {
+            return 1;
+        }
+    } catch {
+        return 0;
+    }
+}
+
+function combineEventSources(previousSearchResult, response, a, b) {
+    // Merge event sources and sort the events.
+    const combinedEvents = a.concat(b).sort(compareEvents);
+    // Put half of the events in the response, and cache the other half.
+    response.results = combinedEvents.slice(0, SEARCH_LIMIT);
+    previousSearchResult.cachedEvents = combinedEvents.slice(SEARCH_LIMIT);
+}
+
+/**
+ * Combine the events from our event sources into a sorted result
+ *
+ * This method will first be called from the combinedSearch() method. In this
+ * case we will fetch SEARCH_LIMIT events from the server and the local index.
+ *
+ * The method will put the SEARCH_LIMIT newest events from the server and the
+ * local index in the results part of the response, the rest will be put in the
+ * cachedEvents field of the previousSearchResult (in this case an empty search
+ * result).
+ *
+ * Every subsequent call will be made from the combinedPagination() method, in
+ * this case we will combine the cachedEvents and the next SEARCH_LIMIT events
+ * from either the server or the local index.
+ *
+ * Since we have two event sources and we need to sort the results by date we
+ * need keep on looking for the oldest event. We are implementing a variation of
+ * a sliding window.
+ *
+ * If we set SEARCH_LIMIT to 3:
+ *
+ *  Server events [01, 02, 04, 06, 07, 08, 11, 13]
+ *                |01, 02, 04|
+ *  Local events  [03, 05, 09, 10, 12, 14, 15, 16]
+ *                |03, 05, 09|
+ *
+ *  We note that the oldest event is from the local index, and we combine the
```

Comment:
Are the numbers meant to convey age / which is oldest, or are they random? At first glance, I expected `01` to mean "first" and thus "oldest", but that seems to disagree with the text that comes after.

Reply:
It's supposed to represent the age, that is `01` is newest and we need to show it as the first search result. By oldest here I meant that in the current sliding windows the oldest (event with the highest number) is from the local event the `09` one.

Reply:
Hmm, okay, doesn't it seem harder to follow that way...? If you ask without other context "which event occurred first, 1 or 2?" I assume many people might say 1 occurred first...?

It also means if a "new" event comes in (not part of your example, but to demonstrate), it would be "newer" than 1, so is it now the "new 1" or is it "0"?

Reply:
It's a bit abstract, it's supposed to represent two sorted lists which we need to merge, which is basically what we get from the server and Seshat.

We can swap it around if you prefer it that way.

Reply:
I guess rather than fiddling with the all the examples to flip things which sounds a bit tedious, I think it would also be clearer if you state the meaning you explained in this thread as part of the example.